<img src="https://www.jobsexpo.ie/wp-content/uploads/2018/09/logo-1030x658.jpg" width = 600>

## Data preperation

- 5 data dicipliner
- Opgave

In [1]:
%%HTML
<style>

.CodeMirror {
    width: 100vw;
}

.container {
    width: 99% !important;
}

.rendered_html {
  font-size:0.8em;
}
.rendered_html table, .rendered_html th, .rendered_html tr, .rendered_html td {
     font-size: 120%;
}

</style>


<img src = "https://www.imgs.ie/ckeditor_uploads/images/DataPreparation%26SelfService(1).jpg">

## Vi skal kigge på 5 discipliner

- **Vælg søjler** Vælg variable / søjler baseret på navn (["navn"]).
- **Tilføj søjler** Lav nye variable som funktion af eksisterende variable.
- **Filtrer** Valg af observationer baseret på deres værdi (where()).
- **Sortér** rækker baseret på værdi (sort_values()).
- **Aggregéring** Kollaps mange værdier til et summary (.agg()).

Vi bruger følgende data

In [45]:
data = ('a,b,c,d\n'
        '1,2,3,4\n'
        '5,6,7,8\n'
        '9,10,11')
from pandas.compat import StringIO

import pandas as pd
df = pd.read_csv(StringIO(data))

In [48]:
df

a   b   c    d
0  1   2   3  4.0
1  5   6   7  8.0
2  9  10  11  NaN

## Vælg søjler

Anvend `["A"]` til at tilgå en søjle (A) i data. og `[["A","B"]]` for at tilgå flere søjler (A,B).


In [47]:
df["a"]

0    1
1    5
2    9
Name: a, dtype: int64

In [49]:
df[["a","b"]]

a   b
0  1   2
1  5   6
2  9  10

## Lav nye søjler

Nye søjler laves blot ved at sætte et *ikke eksisterende søjlenavn* lig de værdier der ønskes i den nye søjle

In [50]:
df2 = df.copy()
df2['e'] = [2,3,4]
df2

a   b   c    d  e
0  1   2   3  4.0  2
1  5   6   7  8.0  3
2  9  10  11  NaN  4

In [52]:
df2['g'] = df2['e']**2
df2

a   b   c    d  e   g
0  1   2   3  4.0  2   4
1  5   6   7  8.0  3   9
2  9  10  11  NaN  4  16

## Filtrer data

Data filtreres ved brug af såkaldte "boolske statements" (Sande eller falske statements).

In [57]:
df.query("b>3 and c<11")

a  b  c    d
1  5  6  7  8.0

In [59]:
df[(df.b>3) & (df.c<11)]

a  b  c    d
1  5  6  7  8.0

## Sortér værdier

Data sorteres ved brug af `sort_values("A")` 

Ønsker man at sortere fra størst til mindst bruges `sort_values("A",ascending=False)`

Har man to søjler man vil sortere på bruges `sort_values(["A","B"])`

Og skal disse ikke blot sortere fra mindst til størst bruges `sort_values(["A","B"], ascending=[False,False])`

In [61]:
df.sort_values("c", ascending=False)

a   b   c    d
2  9  10  11  NaN
1  5   6   7  8.0
0  1   2   3  4.0

## Aggregér

Aggregering besvarer typisk spørgsmål som hvad er gennemsnittet, max, min standard afvigelsen etc. af e.g "A".

In [63]:
df.agg(["min","max"])

a   b   c    d
min  1   2   3  4.0
max  9  10  11  8.0

In [67]:
df.a.agg(["min","max"])

min    1
max    9
Name: a, dtype: int64

## Aggregering sammen med groupby


Denne kombination af funktioner besvarer hvad er den samlede GDP for hvert land eller gennemsnitlig GDP siden 1999 for hvert land. 

In [75]:
avocado = pd.read_csv("https://github.com/christianbomholt/course_data/raw/master/avocado.csv")\
    .drop("Unnamed: 0",axis=1)
avocado.head()

Date  AveragePrice  Total Volume     4046       4225    4770  \
0  2015-12-27          1.33      64236.62  1036.74   54454.85   48.16   
1  2015-12-20          1.35      54876.98   674.28   44638.81   58.33   
2  2015-12-13          0.93     118220.22   794.70  109149.67  130.50   
3  2015-12-06          1.08      78992.15  1132.00   71976.41   72.58   
4  2015-11-29          1.28      51039.60   941.48   43838.39   75.78   

   Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  region  
0     8696.87     8603.62       93.25          0.0  conventional  2015  Albany  
1     9505.56     9408.07       97.49          0.0  conventional  2015  Albany  
2     8145.35     8042.21      103.14          0.0  conventional  2015  Albany  
3     5811.16     5677.40      133.76          0.0  conventional  2015  Albany  
4     6183.95     5986.26      197.69          0.0  conventional  2015  Albany

In [78]:
avocado[["region","AveragePrice"]].groupby("region").mean().head(2)

AveragePrice
region               
Albany       1.561036
Atlanta      1.337959

# Exercise 1.1

PApvin wants to throw a wine-tasting with french wines. Being a stellar host PApvin ofcourse want to ensure that the attendants are only offered wines of reasonable quality within expense range. Help out PApvin make the most of our budget. 

![](https://www.kj.com/sites/default/files/styles/blog_hero_desktop/public/blog/how-to-create-wine-cellar-collection.jpg?itok=lv7wLFop)


- Use `pandas` on the `winemag-data-130k-v2.csv.csv` to:
  - Filter out all non-French wines
  - group by title 
  - create a column with average points
  - print the top 6 wines
  - Are these the wines you would choose? Try to improve top 6
  
